yb-voyager/cmd/importData.go

```diff
@@ -915,54 +896,112 @@ func doOneImport(task *SplitFileImportTask, connPool *tgtdb.ConnectionPool) {
 	}
-	defer conn.Close(context.Background())
 
+	setTargetSchema(conn)
+	return conn
+}
+
+func setTargetSchema(conn *pgx.Conn) {
 	if sourceDBType != POSTGRESQL && target.Schema != YUGABYTEDB_DEFAULT_SCHEMA {
 		setSchemaQuery := fmt.Sprintf("SET SCHEMA '%s'", target.Schema)
 		_, err := conn.Exec(context.Background(), setSchemaQuery)
 		if err != nil {
 			utils.ErrExit("run query %q on target %q: %s", setSchemaQuery, target.Host, err)
 		}
 	}
+}
+
+func dropIdx(conn *pgx.Conn, idxName string) {
+	dropIdxQuery := fmt.Sprintf("DROP INDEX IF EXISTS %s", idxName)
+	log.Infof("Dropping index: %q", dropIdxQuery)
+	_, err := (*conn).Exec(context.Background(), dropIdxQuery)
+	if err != nil {
+		utils.ErrExit("Failed in dropping index: %s", idxName)
+	}
+}
+
+func executeSqlFile(file string, objType string) {
+	log.Infof("Execute SQL file %q on target %q", file, target.Host)
+
+	conn := newTargetConn()
+	defer conn.Close(context.Background())
 
 	var errOccured = 0
-	sqlStrArray := createSqlStrArray(file, "")
-	for _, sqlStr := range sqlStrArray {
-		log.Infof("Run query %q on target %q", sqlStr[1], target.Host)
-		_, err := conn.Exec(context.Background(), sqlStr[0])
+	sqlInfoArr := createSqlStrInfoArray(file, objType)
+	for _, sqlInfo := range sqlInfoArr {
+		if conn == nil {
+			conn = newTargetConn()
+		}
+		err := executeSqlStmtWithRetries(&conn, sqlInfo, objType)
 		if err != nil {
-			log.Errorf("Run query %q on target %q: %s", sqlStr[1], target.Host, err)
-			if strings.Contains(err.Error(), "already exists") {
-				if !target.IgnoreIfExists {
-					fmt.Printf("\b \n    %s\n", err.Error())
-					fmt.Printf("    STATEMENT: %s\n", sqlStr[1])
-					if !target.ContinueOnError {
-						os.Exit(1)
-					}
+			conn.Close(context.Background())
+			conn = nil
+			errOccured = 1
+		}
+	}
+
+	utils.WaitChannel <- errOccured
+	<-utils.WaitChannel
+}
+
+func executeSqlStmtWithRetries(conn **pgx.Conn, sqlInfo sqlInfo, objType string) error {
+	var err error
+	log.Infof("Run query %q on target %q", sqlInfo.formattedStmtStr, target.Host)
+	for retryCount := 0; retryCount <= DDL_MAX_RETRY_COUNT; retryCount++ {
+		_, err = (*conn).Exec(context.Background(), sqlInfo.stmt)
+		if err == nil {
+			return nil
+		}
+
+		log.Errorf("DDL Execution Failed %s", err)
+		if strings.Contains(strings.ToLower(err.Error()), strings.ToLower(SCHEMA_VERSION_MISMATCH_ERR)) &&
+			objType == "INDEX" {
+			// creating fresh connection
+			(*conn).Close(context.Background())
+			*conn = newTargetConn()
+
+			// DROP INDEX in case INVALID index got created
+			dropIdx(*conn, sqlInfo.objName)
+
+			log.Infof("Sleep for 5 seconds before retrying for %dth time", retryCount)
+			time.Sleep(time.Second * 5)
+			log.Infof("RETRYING DDL: %q", sqlInfo.stmt)
+
+			if retryCount == DDL_MAX_RETRY_COUNT {
+				fmt.Printf("\b \n    %s\n", err.Error())
+				fmt.Printf("    STATEMENT: %s\n", sqlInfo.formattedStmtStr)
+				if !target.ContinueOnError {
+					os.Exit(1)
 				}
```

Comment:
This code snippet is duplicated thrice in this function. Move it at the bottom of the function--just before returning `err`. Something like:
```
	for ... {
		_, err = conn.Exec(...)
		if err is retriable {
			// Create new conn.
			// Drop index.
			// Sleep.
            continue
		}
		// Non retriable errors.
		if "already exist?" && arget.IgnoreIfExists {
			err = nil  // Ignore error.
		}
		break

	}
	if err != nil {
		fmt.Printf("\b \n    %s\n", err.Error())
		fmt.Printf("    STATEMENT: %s\n", sqlInfo.formattedStmtStr)
		if !target.ContinueOnError {
			os.Exit(1)
		}
	}
	return err
```

Reply:
@amit-yb soon we are going to move to new import data code path. Is this change really needed, I was thinking of doing something similar but didn't want to add work

Reply:
New import path uses the same code path for importing sequences and indexes. So yes, we need to make this change.

Reply:
ok cool, will do this then